tools/crmadmin.c

```diff
@@ -344,10 +349,62 @@ struct ipc_client_callbacks crm_callbacks = {
     .destroy = crmadmin_ipc_connection_destroy
 };
 
+static void
+ping_callback (pacemakerd_t *pacemakerd,
+               time_t last_good,
+               enum pacemakerd_state state,
+               int rc, gpointer userdata)
+{
+    crm_time_t *crm_when = crm_time_new(NULL);
+    char *pinged_buf = NULL;
+    static int received_responses = 0;
+
+    crm_time_set_timet(crm_when, &last_good);
+    pinged_buf = crm_time_as_string(crm_when,
+        crm_time_log_date | crm_time_log_timeofday | crm_time_log_with_timezone);
+    printf("Status of pacemakerd: %s (%s%s%s)\n",
+        pacemakerd_state_enum2text(state),
+        (rc == pcmk_ok)?"ok":"unclean",
+        (last_good != (time_t) 0)?" @ ":"",
+        (last_good != (time_t) 0)?pinged_buf:"");
+
+    free(pinged_buf);
+    crm_time_free(crm_when);
+
+    if (BE_SILENT && pacemakerd_state_enum2text(state) != NULL) {
+        fprintf(stderr, "%s\n", pacemakerd_state_enum2text(state));
```

Comment:
Yeah that appeared broken to me.
But I wanted it to be consistent with crmd-ping as far as possible and as I wasn't sure who (CTS, RAs, ... haven't investigated) was relying on that behavior I'd like to keep it in sync so that it might be fixed together with crmd-ping that will anyway change with the new API there.

Reply:
That's fine, though I don't think anything relies on it -- crmadmin was created more for debugging than anything else. The man page even calls it a "development tool" and says it is "likely to be replaced by crm_node" (though it's not).